### PR TITLE
t-SNE visualization

### PR DESCRIPTION
**Pull request by AlvieH**
Viz code up on the website, textures not stretching

Important files: App.js, threeVisualizer.js, CoordinatesRetrievalServlet.java, SpritesheetRetrievalServlet.java

https://step-2020-johndallard.uc.r.appspot.com/index.html

### CODE REVIEW

File: react_app/src/App.js [hunk 1]
```diff
@@ -20,11 +30,18 @@ export default class App extends React.Component {
     });
   }
 
+  datasetSelected() {
+    return this.state.selectedDataset !== "";
```
Comment (zkoogle) [inline react_app/src/App.js:34]: return !!this.state.selectedDataset;

File: react_app/src/Form.js [hunk 2]
```diff
@@ -10,14 +10,27 @@ class EmbeddingForm extends React.Component {
   constructor(props) {
     super(props);
     this.state = {
-      values: [],
+      selectedValue: "",
```
Comment (zkoogle) [inline react_app/src/Form.js:13]: Single quotes plz.

File: react_app/src/Form.js [hunk 3]
```diff
@@ -10,14 +10,27 @@ class EmbeddingForm extends React.Component {
...
       createMode: false
     };
 
+    this.setSelection = this.setSelection.bind(this);
```
Comment (zkoogle) [inline react_app/src/Form.js:17]: Just discovered this. Looks like some people use this style but it isn't that readable to me. Can you do:
this.setSelectionFn_ = this.setSelection.bind(this)
Instead? and use setSelectionFn?
@askewc WDYT?

File: react_app/src/threeVisualizer.js [hunk 4]
```diff
@@ -0,0 +1,230 @@
+import * as THREE from 'three';
```
Comment (zkoogle) [inline react_app/src/threeVisualizer.js:1]: The module name should be lowerCamelCase:
https://google.github.io/styleguide/jsguide.html#es-module-imports

File: react_app/src/threeVisualizer.js [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    this.canvas = undefined;
+    this.light = undefined;
+  }
+
+  // Loads json object for dataset and returns array of points
+  async initPoints(material) {
+    var numImages = 0;
+    const numRows = 1;
+
+    const json_url = "/coordinates-retrieval?dataset=".concat(this.dataset);
```
Comment (zkoogle) [inline react_app/src/threeVisualizer.js:139]: singe quotes

File: src/main/java/com/google/sps/servlets/CoordinatesRetrievalServlet.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    UserService userService = UserServiceFactory.getUserService();
+    String userEmail = userService.getCurrentUser().getEmail();
+    
+    // Get the name of the dataset they want to visualize
+    String datasetName = request.getParameter("dataset"); 
+
+    // Get the path to the spritesheet
+    String userDirectory = BUCKET_NAME + "/" + userEmail + "/";
+    String userDatasetDir = userDirectory + datasetName + "/";
+    String userJsonFilePath = userDatasetDir + "coordinates.json";
```
Comment (zkoogle) [inline src/main/java/com/google/sps/servlets/CoordinatesRetrievalServlet.java:62]: How about creating something like and using it:
  public static String joinPaths(String... paths) {
    return Joiner.on(File.separator).join(paths);
  }
("Joiner" is in Guava).

File: src/main/java/com/google/sps/servlets/CoordinatesRetrievalServlet.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+    log.info("OH HAI THAR! Downloading: " + userJsonFilePath);
+
+    try {
+      BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+      BlobKey blobKey = blobstoreService.createGsBlobKey("/gs/" + userJsonFilePath);
+      blobstoreService.serve(blobKey, response);
+      log.info("KTHNXBAI" + userJsonFilePath);
```
Comment (zkoogle) [inline src/main/java/com/google/sps/servlets/CoordinatesRetrievalServlet.java:70]: What is KTHNXBAI?

File: src/main/java/com/google/sps/servlets/SpritesheetRetrievalServlet.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    String userDirectory = BUCKET_NAME + "/" + userEmail + "/";
+    String userDatasetDir = userDirectory + datasetName + "/";
+    String userSpritesheetPath = userDatasetDir + "spritesheets/spritesheet.png";
+    
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    BlobKey blobKey = blobstoreService.createGsBlobKey(
+        "/gs/" + userSpritesheetPath);
+    blobstoreService.serve(blobKey, response);
+  }
+}
```
Comment (zkoogle) [inline src/main/java/com/google/sps/servlets/SpritesheetRetrievalServlet.java:66]: Please add an extra line.

File: react_app/src/App.js (hunk 1 above)
Comment (AlvieH) [reply react_app/src/App.js]: Fixed!

File: react_app/src/Form.js (hunk 2 above)
Comment (AlvieH) [reply react_app/src/Form.js]: Fixed

File: react_app/src/Form.js (hunk 3 above)
Comment (AlvieH) [reply react_app/src/Form.js]: Fixed!

File: react_app/src/threeVisualizer.js (hunk 4 above)
Comment (AlvieH) [reply react_app/src/threeVisualizer.js]: Fixed!

File: react_app/src/threeVisualizer.js (hunk 5 above)
Comment (AlvieH) [reply react_app/src/threeVisualizer.js]: aaaaaand fixed

File: src/main/java/com/google/sps/servlets/CoordinatesRetrievalServlet.java (hunk 6 above)
Comment (AlvieH) [reply src/main/java/com/google/sps/servlets/CoordinatesRetrievalServlet.java:62]: Since this is @jdstriker731's code I haven't made changes to the java files - equivalent servlets should be in his pr

File: src/main/java/com/google/sps/servlets/CoordinatesRetrievalServlet.java (hunk 7 above)
Comment (AlvieH) [reply src/main/java/com/google/sps/servlets/CoordinatesRetrievalServlet.java:70]: I think it's short for "Okay thanks, bye"

@jdstriker731 pls comment

File: src/main/java/com/google/sps/servlets/SpritesheetRetrievalServlet.java (hunk 8 above)
Comment (AlvieH) [reply src/main/java/com/google/sps/servlets/SpritesheetRetrievalServlet.java:66]: @jdstriker731

File: pom.xml [hunk 9]
```diff
@@ -61,7 +61,7 @@
         <version>2.2.0</version>
         <configuration>
           <!-- TODO: set project ID. -->
-          <!-- <deploy.projectId>[PROJECT ID HERE]</deploy.projectId> -->
```
Comment (zkoogle) [inline pom.xml:64]: nit: Please revert this change.

File: react_app/src/App.js [hunk 10]
```diff
@@ -1,14 +1,24 @@
...
 import EmbeddingForm from './Form';
 import styles from './App.scss';
+import {ThreeRenderer} from './threeVisualizer.js'
 
 export default class App extends React.Component {
   constructor(props) {
     super(props);
-    this.state = { datasets: [] };
+    this.state = { datasets: [],
+                    selectedDataset: "" };
```
Comment (zkoogle) [inline react_app/src/App.js:10]: single quotes, here and in all the files.

File: react_app/src/threeVisualizer.js [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+    this.width = image.width * numColumns;
+    this.height = image.height * numRows;
+    this.numImages = numImages;
+  }
+}
+
+class Point {
+  constructor(x, y, z) {
+    function randomInt() {
+      const val = Math.random() * 700;
```
Comment (zkoogle) [inline react_app/src/threeVisualizer.js:26]: Where did "700" or "-400" come from? Shouldn't they be calculated values from some scale param?
At least take them out as constants.

File: react_app/src/threeVisualizer.js [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+  constructor(image, numImages, numRows, numColumns) {
+    this.image = image;
+    this.width = image.width * numColumns;
+    this.height = image.height * numRows;
+    this.numImages = numImages;
+  }
+}
+
+class Point {
+  constructor(x, y, z) {
```
Comment (zkoogle) [inline react_app/src/threeVisualizer.js:24]: This constructor deserves some comments. What is the usecase of the random assignment?

File: react_app/src/threeVisualizer.js [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+const IMAGE_SIZE = 64;
+
+class ImageParams {
+  constructor(size) {
+    this.width = size;
+    this.height = size;
+  }
+}
+
+class Atlas {
```
Comment (zkoogle) [inline react_app/src/threeVisualizer.js:14]: Can you add a class comment?

File: react_app/src/threeVisualizer.js [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+
+class AtlasGeometry extends three.Geometry {
+  constructor(atlas, points) {
+    super();
+
+    for (var i = 0; i < atlas.numImages; i++) {
+      // Create x, y, z coords for this subimage
+      const coords = points[i];
+
+      this.vertices.push(
```
Comment (zkoogle) [inline react_app/src/threeVisualizer.js:58]: 1. Can you add a comment? e.g., "3D vertices of the image thumbnails"
2. Technically, image thumbnail is NOT in the center of the coords but it is not a problem because the image sizes are the same (coordinate translation by the half of the image width/height). You may write that down in the comment, too.

File: react_app/src/threeVisualizer.js [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+    const atlas = new Atlas(new ImageParams(IMAGE_SIZE), numImages, numRows, numColumns);
+
+    const geometry = new AtlasGeometry(atlas, points);
+
+    const mesh = new three.Mesh(geometry, this.material);
+    mesh.position.set(0, 0, 0);
+    this.scene.add(mesh);
+
+    const component = this;
+    function render3d() {
```
Comment (zkoogle) [inline react_app/src/threeVisualizer.js:160]: Why do you need to create a nested function for this?

File: react_app/src/App.js (hunk 10 above)
Comment (AlvieH) [reply react_app/src/App.js]: Fixed!

File: react_app/src/threeVisualizer.js (hunk 11 above)
Comment (AlvieH) [reply react_app/src/threeVisualizer.js]: Factored them out to be comments - default construction isn't used in the final product, and it's mostly useful for debugging

File: react_app/src/threeVisualizer.js (hunk 12 above)
Comment (AlvieH) [reply react_app/src/threeVisualizer.js:33]: Added class comments to all classes

File: react_app/src/threeVisualizer.js (hunk 13 above)
Comment (AlvieH) [reply react_app/src/threeVisualizer.js:19]: Added class comments to all classes

File: react_app/src/threeVisualizer.js (hunk 14 above)
Comment (AlvieH) [reply react_app/src/threeVisualizer.js:72]: Added

File: react_app/src/threeVisualizer.js (hunk 15 above)
Comment (AlvieH) [reply react_app/src/threeVisualizer.js:176]: In this case it worked the best, @askewc and I worked on this for a bit and it seems to introduce rendering issues if it's not a nested function

File: pom.xml (hunk 9 above)
Comment (AlvieH) [reply pom.xml:64]: Done!

File: react_app/src/threeVisualizer.js (hunk 15 above)
Comment (zkoogle) [reply react_app/src/threeVisualizer.js:176]: Oh.. because of the "requestAnimationFrame" which calls the self. That's fine with me (although there should be a better way of handling it; such as taking it out as an external function and adding a proper binding). Can you at least add a comment?

File: react_app/src/App.js [hunk 16]
```diff
@@ -1,30 +1,47 @@
...
 import EmbeddingForm from './Form';
 import styles from './App.scss';
+import {ThreeRenderer} from './threeVisualizer.js'
 
 export default class App extends React.Component {
   constructor(props) {
     super(props);
-    this.state = { datasets: [] };
+    this.state = { datasets: [],
+                    selectedDataset:  '' };
```
[the remark this answers is not in the document]
Comment (AlvieH) [reply react_app/src/App.js]: Fixed!

File: react_app/src/threeVisualizer.js (hunk 12 above)
Comment (AlvieH) [reply react_app/src/threeVisualizer.js:33]: Should be commented there in newest commit, weird it wasn't showing up for you

File: pom.xml (hunk 9 above)
Comment (AlvieH) [reply pom.xml:64]: It's showing as done for me? Weird

File: src/main/java/com/google/sps/servlets/SpritesheetRetrievalServlet.java (hunk 8 above)
Comment (jdstriker731) [reply src/main/java/com/google/sps/servlets/SpritesheetRetrievalServlet.java:66]: @zkoogle where do you mean? at the end of the file?

File: pom.xml (hunk 9 above)
Comment (zkoogle) [reply pom.xml:64]: This is what I see:
https://screenshot.googleplex.com/F02oDCv5tA3.png
But it's not important enough to block submission, tho.

File: react_app/src/threeVisualizer.js (hunk 15 above)
Comment (AlvieH) [reply react_app/src/threeVisualizer.js]: Factored into member function with binding

File: react_app/src/threeVisualizer.js (hunk 12 above)
Comment (zkoogle) [reply react_app/src/threeVisualizer.js:33]: My question is when would someone call this with the values not present? Why do you assign them with random values (for what purpose)?

File: src/main/java/com/google/sps/servlets/SpritesheetRetrievalServlet.java (hunk 8 above)
Comment (zkoogle) [reply src/main/java/com/google/sps/servlets/SpritesheetRetrievalServlet.java:66]: Yes, the end of the file should end with a blank line (to remove that red circle).

File: pom.xml (hunk 9 above)
Comment (AlvieH) [reply pom.xml:64]: Ohhh I misunderstood the request, fixed!

Comment (AlvieH) [reply pom.xml:64]: Ohhh my bad I misunderstood the request, fixed in latest commit

File: react_app/src/threeVisualizer.js (hunk 12 above)
Comment (AlvieH) [reply react_app/src/threeVisualizer.js:33]: It was useful earlier in the process (before there was actual data) for debugging purposes

Comment (AlvieH) [reply react_app/src/threeVisualizer.js:33]: I can remove if you'd prefer, but imo it's not doing damage being there

Comment (zkoogle) [reply react_app/src/threeVisualizer.js:33]: Please remove it. It will make maintenance harder and it can potentially introduce error/bug to someone who misuses it.

Comment (AlvieH) [reply react_app/src/threeVisualizer.js:28]: Removed!

File: src/main/java/com/google/sps/servlets/CoordinatesRetrievalServlet.java (hunk 7 above)
Comment (jdstriker731) [reply src/main/java/com/google/sps/servlets/CoordinatesRetrievalServlet.java]: I'm confused. Do we still need this or was it just for your own testsing @AlvieH ?
